src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -37,9 +69,63 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 using (var memoryStream = new MemoryStream(value))
                 using (var writer = new BsonReader(memoryStream))
                 {
-                    tempDataDictionary = jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                    tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
                 }
 
+                var convertedDictionary = new Dictionary<string, object>(tempDataDictionary, StringComparer.OrdinalIgnoreCase);
+                foreach (var item in tempDataDictionary)
+                {
+                    var jArrayValue = item.Value as JArray;
+                    if (jArrayValue != null && jArrayValue.Count > 0)
+                    {
+                        var arrayType = jArrayValue[0].Type;
+                        Type returnType;
+                        if (_tokenTypeLookup.TryGetValue(arrayType, out returnType))
+                        {
+                            var arrayConverter = _arrayConverters.GetOrAdd(returnType, type =>
+                            {
+                                return (Func<JArray, object>)_convertArrayMethodInfo.MakeGenericMethod(type).CreateDelegate(typeof(Func<JArray, object>));
+                            });
+                            var result = arrayConverter(jArrayValue);
+
+                            convertedDictionary[item.Key] = result;
+                        }
+                        else
+                        {
+                            var message = Resources.FormatTempData_CannotDeserializeToken(arrayType);
+                            throw new InvalidOperationException(message);
+                        }
+                    }
+                    else
+                    {
+                        var jObjectValue = item.Value as JObject;
+                        if (jObjectValue == null || !jObjectValue.HasValues)
+                        {
+                            continue;
```

Comment:
Should we assign a null value when `jObjectValue.HasValues == false`? It will avoid throwing a cast exception in user code:

```C#
TempData["key"] = new Dictionary<string, string>();

...
var myDictionary = (IDictionary<string, string>)TempData["key"]; // will throw now
```